src/uct/ib/dc/dc_mlx5_ep.c

```diff
@@ -588,77 +588,115 @@ ucs_status_t uct_dc_mlx5_ep_get_zcopy(uct_ep_h tl_ep, const uct_iov_t *iov,
     return UCS_INPROGRESS;
 }
 
-ucs_status_t uct_dc_mlx5_ep_flush(uct_ep_h tl_ep, unsigned flags,
-                                  uct_completion_t *comp)
+static UCS_F_ALWAYS_INLINE ucs_status_t
+uct_dc_mlx5_ep_flush_cancel(uct_dc_mlx5_ep_t *ep, uint16_t *sn_p)
 {
-    uct_dc_mlx5_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_dc_mlx5_iface_t);
-    uct_dc_mlx5_ep_t    *ep    = ucs_derived_of(tl_ep, uct_dc_mlx5_ep_t);
-    uct_ib_mlx5_md_t    *md    = ucs_derived_of(iface->super.super.super.super.md,
-                                                uct_ib_mlx5_md_t);
-    uint8_t pool_index         = uct_dc_mlx5_ep_pool_index(ep);
-    ucs_status_t        status;
-    uint16_t            sn;
+    uct_dc_mlx5_iface_t *iface = ucs_derived_of(ep->super.super.iface,
+                                                uct_dc_mlx5_iface_t);
+    uct_ib_mlx5_md_t *md       =
+            ucs_derived_of(iface->super.super.super.super.md,
+                           uct_ib_mlx5_md_t);
+    ucs_status_t status;
     UCT_DC_MLX5_TXQP_DECL(txqp, txwq);
 
-    if (ep->dci == UCT_DC_MLX5_EP_NO_DCI) {
-        if (uct_dc_mlx5_iface_dci_can_alloc(iface, pool_index)) {
-            UCT_TL_EP_STAT_FLUSH(&ep->super); /* no sends */
-            return UCS_OK;
-        }
-
-        return UCS_ERR_NO_RESOURCE; /* waiting for dci */
+    if (uct_dc_mlx5_iface_is_dci_rand(iface)) {
+        return UCS_ERR_UNSUPPORTED;
     }
 
-    if (!uct_dc_mlx5_iface_has_tx_resources(iface)) {
-        return UCS_ERR_NO_RESOURCE;
+    if (ep->dci == UCT_DC_MLX5_EP_NO_DCI) {
+        UCT_TL_EP_STAT_FLUSH(&ep->super);
+        return UCS_OK; /* nothing to cancel */
     }
 
-    if (!uct_dc_mlx5_iface_dci_ep_can_send(ep)) {
+    if (!uct_dc_mlx5_iface_has_tx_resources(iface) ||
```

Comment:
why need to check `uct_dc_mlx5_iface_has_tx_resources`?

Reply:
to be protected against CQ overrun

Reply:
it is not possible with dc, please check what `uct_dc_mlx5_iface_has_tx_resources` does

Reply:
> it is not possible with dc, please check what uct_dc_mlx5_iface_has_tx_resources does

thanks, fixed